### PR TITLE
Add PCS /page/media endpoint

### PR DESCRIPTION
**Pull request by berndsi**
Simple call to PCS/MCS, no storage at this time.

https://phabricator.wikimedia.org/T189751

### CODE REVIEW

File: projects/dev.yaml [hunk 1]
```diff
@@ -25,6 +25,10 @@ paths:
                 - path: v1/mobileapps.yaml
                   options: '{{merge({"response_cache_control": options.purged_cache_control},
                                 options.mobileapps)}}'
+                - path: v1/media.yaml
+                  options:
+                    response_cache_control: '{{options.purged_cache_control}}'
```
Comment (Pchelolo) [inline projects/dev.yaml]: If this is purged we need a ChangeProp rule to purge it. I'll make one.

File: v1/media.yaml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    name: Reading Infrastructure
+    url: https://www.mediawiki.org/wiki/Wikimedia_Reading_Infrastructure_team
+  license:
+    name: Apache licence, v2
+    url: https://www.apache.org/licenses/LICENSE-2.0
+paths:
+  /media/{title}{/revision}:
+    get:
+      tags:
+        - Mobile
```
Comment (Pchelolo) [inline v1/media.yaml]: Is it indeed mobile? This seems useful elsewhere.. Maybe put it under pagecontent?

File: v1/media.yaml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+              body: '{{get_from_pcs.body}}'
+
+#      x-monitor: true
+
+      x-amples:
+        - title: Get media in Foobar page
+          request:
+            params:
+              domain: en.wikipedia.org
+              title: Foobar
```
Comment (Pchelolo) [inline v1/media.yaml]: We try to point it to some pages on test wiki or in user space to avoid breakage from use edits.

File: v1/media.yaml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+              headers:
+                cache-control: '{{cache-control}}'
+              uri: '{{$$.options.host}}/{domain}/v1/page/media/{title}/{revision}'
+        - return_to_client:
+            return:
+              status: 200
+              headers: '{{get_from_pcs.headers}}'
+              body: '{{get_from_pcs.body}}'
+
+#      x-monitor: true
```
Comment (Pchelolo) [inline v1/media.yaml]: You still need this.

File: v1/media.yaml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            return HTTP 302 with a redirect target in `Location` header and content in the body.
+
+            To get a 200 response instead, supply `false` to the `redirect` parameter.
+          type: boolean
+          required: false
+      responses:
+        '200':
+          description: JSON containing metadata of media items appearing on the given page.
+          schema:
+            type: object
```
Comment (Pchelolo) [inline v1/media.yaml]: Is there any known schema we could specify?

File: v1/media.yaml (hunk 4 above)
Comment (berndsi) [reply v1/media.yaml]: Will set to true. I assume this is for Icinga monitoring, and using the x-ample below.

File: v1/media.yaml (hunk 5 above)
Comment (berndsi) [reply v1/media.yaml]: Yes, will add.

File: projects/dev.yaml (hunk 1 above)
Comment (berndsi) [reply projects/dev.yaml]: Yes, the other Page Content Service endpoints (except the base CSS one) will need the same.

File: v1/media.yaml (hunk 2 above)
Comment (berndsi) [reply v1/media.yaml]: Ok, will move to pagecontent.

**Comment by berndsi** [comment]
@Pchelolo Any ideas why the Etag emitted by RB starts with `W/`? The ETag from MCS doesn't.

E.g. compare the ETag from
http://localhost:7231/en.wikipedia.org/v1/page/media/Foobar/828405720
with
http://appservice.wmflabs.org/en.wikipedia.org/v1/page/media/Foobar/828405720

**Comment by berndsi** [comment]
@Pchelolo Why/how does RB add a `W/` in front of the Etag value?

Compare 
```
curl -sI "http://localhost:7231/en.wikipedia.org/v1/page/media/Foobar/828405720" | grep -i etag
etag: W/"828405720/96d053c7-1e02-11e8-858d-37608f62c181"
```
with
```
curl -sI "http://appservice.wmflabs.org/en.wikipedia.org/v1/page/media/Foobar/828405720" | grep -i etag
etag: "828405720/96d053c7-1e02-11e8-858d-37608f62c181"
```

**Comment by Pchelolo** [comment]
@berndsi doesn't happen for me... Usually the `W/` - an indication of a weak etag is added by Varnish. Are you using the appservice.wmflabs.org in your config.yaml for mobileapps host?

**Comment by berndsi** [comment]
@Pchelolo Yes, I'm using appservice.wmflabs.org in my config.yaml for mobileapps host. Same seems to happen in the test. That's one of the test failures in CI (besides the Parsoid and the PDF ones).

**Comment by Pchelolo** [comment]
@berndsi for CI please rebase on current master, I've just fixed some unrelated test failures.

Re weak etag - I'm surprised. I've downloaded your patch and tried myself and I don't get the weak etag.. weeeeird.

**Comment by berndsi** [comment]
@Pchelolo works for me locally at least since the rebase.

**Comment by berndsi** [comment]
@Pchelolo Are the Node 8 tests more flaky than the Node 6 tests or is that a coincidence? Is there something like a Gerrit recheck here?

**Comment by Pchelolo** [comment]
> Are the Node 8 tests more flaky than the Node 6 tests or is that a coincidence? Is there something like a Gerrit recheck here?

Ye, you click on Travis check and restart the build

File: config.test.yaml [hunk 6]
```diff
@@ -8,7 +8,7 @@ default_project: &default_project
     - path: projects/wmf_default.yaml
       options: &default_options
         table:
-          backend: cassandra
+          backend: sqlite
```
Comment (Pchelolo) [inline config.test.yaml]: This is clearly from some rebasing weirdness or github being buggy..

Could you try to rebase again and force-push into your branch? 

like on master `git pull upstream master; git checkout pcs/media; git rebase master; git push -f origin pcs/media`

File: projects/dev.yaml (hunk 1 above)
Comment (Pchelolo) [reply projects/dev.yaml]: I guess it's fine to assume then whenever Parsoid HTML changes -> media might also change so we need a rule for parsoid HTML change?

File: v1/media.yaml [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    url: https://www.apache.org/licenses/LICENSE-2.0
+paths:
+  /media/{title}{/revision}:
+    x-route-filters:
+      - path: ./lib/revision_table_access_check_filter.js
+        options:
+          redirect_cache_control: '{{options.response_cache_control}}'
+      - path: lib/security_response_header_filter.js
+        options:
+          allowInlineStyles: true
```
Comment (Pchelolo) [inline v1/media.yaml]: We don't need this - it's JSON, not HTML, so no inline styles needed

File: config.test.yaml (hunk 6 above)
Comment (berndsi) [reply config.test.yaml]: Ok, rebased. It's weird that I have to do a `-f`.

File: v1/media.yaml (hunk 7 above)
Comment (berndsi) [reply v1/media.yaml]: Done. Question: What exactly does the `redirect_cache_control: '{{options.response_cache_control}}'` on line 18 do?

Comment (Pchelolo) [reply v1/media.yaml]: That's the cache-control header that will be applied to 301/302 responces

**Comment by berndsi** [comment]
@Pchelolo @d00rman Should we add cache-control for this here or in MCS? What values do you suggest? I was thinking we should have a bit of caching since it's not stored, maybe set it to 5 minutes client side and a long period for Varnish. We'll still have the purging stuff, right?

File: v1/media.yaml [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+            $ref: '#/definitions/problem'
+
+      x-request-handler:
+        - get_from_pcs:
+            request:
+              method: get
+              headers:
+                cache-control: '{{cache-control}}'
+              uri: '{{$$.options.host}}/{domain}/v1/page/media/{title}/{revision}'
+        - return_to_client:
```
Comment (d00rman) [inline v1/media.yaml]: This is not needed since the response is not being modifed

File: v1/media.yaml [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            return:
+              status: 200
+              headers: '{{get_from_pcs.headers}}'
+              body: '{{get_from_pcs.body}}'
+
+      x-monitor: true
+
```
Comment (d00rman) [inline v1/media.yaml]: Remove the empty line

File: v1/media.yaml [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+          schema:
+            $ref: '#/definitions/problem'
+
+      x-request-handler:
+        - get_from_pcs:
+            request:
+              method: get
+              headers:
+                cache-control: '{{cache-control}}'
+              uri: '{{$$.options.host}}/{domain}/v1/page/media/{title}/{revision}'
```
Comment (d00rman) [inline v1/media.yaml]: `options.host` is enough

**Comment by Pchelolo** [comment]
@berndsi for mobileapps endpoints we add the header in restbase, so let's do the same. Here's the example: https://github.com/wikimedia/restbase/blob/master/v1/mobileapps.yaml#L88

**Comment by Pchelolo** [comment]
Made some benchmarking:

```
ab -c 10 -n 100 http://mobileapps.svc.codfw.wmnet:8888/en.wikipedia.org/v1/page/media/Barack_Obama

Concurrency Level:      10
Time taken for tests:   19.425 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      10906700 bytes
HTML transferred:       10818200 bytes
Requests per second:    5.15 [#/sec] (mean)
Time per request:       1942.543 [ms] (mean)
Time per request:       194.254 [ms] (mean, across all concurrent requests)
Transfer rate:          548.31 [Kbytes/sec] received
```

```
ab -c 10 -n 100 http://mobileapps.svc.codfw.wmnet:8888/en.wikipedia.org/v1/page/media/Russia

Concurrency Level:      10
Time taken for tests:   18.794 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      18715500 bytes
HTML transferred:       18627000 bytes
Requests per second:    5.32 [#/sec] (mean)
Time per request:       1879.434 [ms] (mean)
Time per request:       187.943 [ms] (mean, across all concurrent requests)
Transfer rate:          972.47 [Kbytes/sec] received
```

So requests are indeed pretty slow and we will need to have storage.

File: v1/media.yaml (hunk 8 above)
Comment (berndsi) [reply v1/media.yaml]: @d00rman You mean `uri` or `headers` or both?

Comment (d00rman) [reply v1/media.yaml]: Neither, I'm referring to the `return_to_client` block - it's not needed at all, so you can remove it.

Comment (Pchelolo) [reply v1/media.yaml]: @d00rman it will be needed as we want to add the cache-control header.

Comment (berndsi) [reply v1/media.yaml]: @d00rman Got it. But this might change if we need to have storage (see @Pchelolo 's test run below).

File: v1/media.yaml [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+        - return_to_client:
+            return:
+              status: 200
+              headers: '{{get_from_pcs.headers}}'
```
Comment (Pchelolo) [inline v1/media.yaml]: ok, just add cache-control here like in https://github.com/wikimedia/restbase/blob/master/v1/mobileapps.yaml#L88 and imho we are good to go

File: v1/media.yaml (hunk 9 above)
Comment (berndsi) [reply v1/media.yaml]: OK, removing some more empty lines for consistency.

File: v1/media.yaml (hunk 8 above)
Comment (berndsi) [reply v1/media.yaml]: Done.

**Comment by Pchelolo** [comment]
👍 @d00rman you to?

**Comment by Pchelolo** [comment]
As usual PDF failed, but it's not related. Merging.

**Comment by mdholloway** [comment]
Hey @Pchelolo, where did you run the above benchmarking?  I tried from scb1001 but it doesn't have `ab` installed.

**Comment by Pchelolo** [comment]
restbase-dev. Do you have shell access there? I don;'t think so..

**Comment by mdholloway** [comment]
Nope.

**Comment by Pchelolo** [comment]
hm... I don't think we have AB installed anywhere in production. You could try creating a task and ask ops about that.

In the meantime you can give me a set of commands and I'll execute them. use mobileapps.svc.eqiad.wmnet for MCS host

**Comment by mdholloway** [comment]
I was just going to run the exact same queries as you did before. 

`ab -c 10 -n 100 http://mobileapps.svc.codfw.wmnet:8888/en.wikipedia.org/v1/page/media/Barack_Obama`

`ab -c 10 -n 100 http://mobileapps.svc.codfw.wmnet:8888/en.wikipedia.org/v1/page/media/Russia`

**Comment by Pchelolo** [comment]
Ok, apparently I was running them in codfw from restbase-test that doesn't exist any more... I can run it from eqiad, but results might be a bit convoluted because of that. Anyway:

```
ab -c 10 -n 100 http://mobileapps.svc.codfw.wmnet:8888/en.wikipedia.org/v1/page/media/Barack_Obama

Server Software:        
Server Hostname:        mobileapps.svc.eqiad.wmnet
Server Port:            8888

Document Path:          /en.wikipedia.org/v1/page/media/Barack_Obama
Document Length:        123870 bytes

Concurrency Level:      10
Time taken for tests:   22.876 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      12477700 bytes
HTML transferred:       12387000 bytes
Requests per second:    4.37 [#/sec] (mean)
Time per request:       2287.626 [ms] (mean)
Time per request:       228.763 [ms] (mean, across all concurrent requests)
Transfer rate:          532.66 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       1
Processing:  1577 2264 456.7   2164    3678
Waiting:     1575 2261 457.5   2160    3676
Total:       1577 2264 456.8   2165    3679

Percentage of the requests served within a certain time (ms)
  50%   2165
  66%   2325
  75%   2445
  80%   2532
  90%   3172
  95%   3354
  98%   3484
  99%   3679
 100%   3679 (longest request)

ab -c 10 -n 100 http://mobileapps.svc.eqiad.wmnet:8888/en.wikipedia.org/v1/page/media/Russia

Concurrency Level:      10
Time taken for tests:   41.912 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      20780400 bytes
HTML transferred:       20689700 bytes
Requests per second:    2.39 [#/sec] (mean)
Time per request:       4191.236 [ms] (mean)
Time per request:       419.124 [ms] (mean, across all concurrent requests)
Transfer rate:          484.19 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0       4
Processing:  3618 4110 237.4   4090    4755
Waiting:     3615 4106 237.1   4087    4752
Total:       3618 4110 237.4   4090    4755

Percentage of the requests served within a certain time (ms)
  50%   4090
  66%   4212
  75%   4280
  80%   4311
  90%   4388
  95%   4556
  98%   4701
  99%   4755
 100%   4755 (longest request)

```

**Comment by Pchelolo** [comment]
Seems like it became worse :(

**Comment by mdholloway** [comment]
Ah, right: the new endpoint requires query continuation of the MW API queries for imageinfo in the case of very large pages (definitely for Barack Obama, probably for Russia as well).  In our local benchmarking, the big performance gains were for medium-size pages like Hummingbird.  Unfortunately I don't think we have any historical data for those.

**Comment by mdholloway** [comment]
Meanwhile, the internal p75 for the past 24 hours is basically flat:

https://grafana.wikimedia.org/dashboard/db/mobileapps?panelId=9&fullscreen&orgId=1&from=now-24h&to=now

(The change was deployed less than 24 hours ago, last night U.S. time.)

**Pull request merged**